### DISCUSSION TITLE
MassTransit Consumer Testing

### DISCUSSION THREAD

**Discussion by kron0s19** [jasontaylordev/CleanArchitecture#1079] [Q&A] [answered]
Hey ! 

Im having issues trying to test an **IConsumer** using **MassTransit**. Everything works properly, but I cannot figure out how to create an integration test to check that the consumer actually gets fired.

The problem is that I cannot make the **TestHarness** "wait" until the consumer gets fired. It always gets executed after the test and that's why It fails. I can create a loop and ** await Task.Delay(100);** to wait for it and it will eventually work. But It is not what I was expecting. 

Any help would be greatly appreciated! 

In my web app factory:
```
protected override void ConfigureWebHost(IWebHostBuilder builder)
{
    builder
        .UseConfiguration(_configuration)
        .ConfigureAppConfiguration(configBuilder => configBuilder.AddConfiguration(_configuration))
        .ConfigureTestServices(services =>
        {
            services.AddMassTransitTestHarness(x =>
            {
                x.AddConsumer<NoteCreatedConsumer>();
            });
}
```

And this is my test

```
[Fact]
public async Task ShouldCreateNote_WhenScribeNoteCreatedIntegrationEvent()
{
    // Arrange
    NoteCreatedIntegrationEvent @event = new();

    // Act
    using var scope = GetScopeFactory().CreateScope();
    var testHarness = scope.ServiceProvider.GetRequiredService<ITestHarness>();
    await testHarness.Start();

    // Option1: publish event using testHarness.Bus.    
    await testHarness.Bus.Publish(@event);

    //Option2: Use the IEventBus interface to publish the event (this is what it gets used in the app)
    //var eventBus = scope.ServiceProvider.GetRequiredService<IEventsBus>();
    //await eventBus.Publish(@event);

    var consumerTestHarness = testHarness.GetConsumerHarness<NoteCreatedConsumer>();
    var result = await consumerTestHarness.Consumed.Any<NoteCreatedIntegrationEvent>();

   

    // Assert       
    // ****** THIS FAILS************  it it always false**
     //result.Should().BeTrue();

    // ****** WORKAROUND************ This loop should not be necesary: I want to remove it.**
    while (!await consumerTestHarness.Consumed.Any<ScribeNoteCreatedIntegrationEvent>())
    {
        await Task.Delay(100);
    }
    // my stuff..
}
```

**Comment by campbellja** [comment, marked as answer]
Hi @kron0s19 If this helps, you can wait until the bus has finished processing your event by awaiting the InactivityTask
> await testHarness.InactivityTask;

Please See Chris P's comment on a similar question to yours: https://github.com/MassTransit/MassTransit/discussions/4890#discussioncomment-8182634


Also the Test Harness's `Published` property contains a list which you can use to assert if it has been published or not: `testHarness.Published...`

**Reply by kron0s19** [reply]
It actually works with the **_InactiviyTask_**. Thanks for the reply mate!
